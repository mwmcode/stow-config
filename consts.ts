export const TestDirectories = {
  configSrcDir: 'test/data/config-src',
  configDestDir: 'test/data/config-dest',
};

export const TESTSC_HOME_ALTERNATIVE = 'TESTSC_CONFIG';

export const TEST_DENO_ARGS = [
  '-R',
  `--allow-env=SNAPSHOT_TEST_NAME,${TESTSC_HOME_ALTERNATIVE}`,
];
