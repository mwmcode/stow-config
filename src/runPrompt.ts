import { normalize } from '@std/path';
import { Checkbox, Input, prompt, Toggle } from '@cliffy/prompt';
import type { CheckboxOption } from '@cliffy/prompt';
import { validateDirInput } from './validateDirExists.ts';
import type { PromptResult } from './types.ts';
import { logger } from './logger.ts';

export async function runPrompt() {
  const configFiles: CheckboxOption<string>[] = [];
  let hasCreatedDestDir = false;

  return (await prompt([
    {
      name: 'configDir',
      type: Input,
      files: true,
      message: logger.prompt('Enter path to `configs` directory'),
      transform: normalize,
      after: async ({ configDir }, next) => {
        configDir = configDir as string;
        const configsDirFullpath = normalize(configDir);
        const dirValidatoin = validateDirInput(configDir);
        if (dirValidatoin === 'is-directory') {
          for await (
            const { isSymlink, name } of Deno.readDir(
              configsDirFullpath,
            )
          ) {
            if (isSymlink) continue;
            configFiles.push({ name, value: name });
          }
          if (!configFiles.length) {
            logger.error(`${logger.underline(configDir)} contains no configs!`);
            await next('configDir');
          }
          logger.emptyLine();
          await next();
        } else if (['is-new', 'is-file'].includes(dirValidatoin)) {
          logger.error(
            `No such directory (${
              logger.underline(
                configDir,
              )
            }). Please try again`,
          );
          await next('configDir');
        }
      },
    },
    {
      name: 'destDir',
      type: Input,
      files: true,
      message: logger.prompt('Enter path to `destination` directory'),
      transform: normalize,
      after: async ({ destDir, configDir }, next) => {
        const dirValidation = validateDirInput(destDir as string);
        if (destDir === configDir) {
          logger.error(
            `${
              logger.underline(
                destDir,
              )
            } cannot be the same as configs (source) directory`,
          );
          await next('destDir');
        }
        if (dirValidation === 'is-file') {
          logger.error(`${logger.underline(destDir)} must be a directory`);
          await next('destDir');
        }
        if (dirValidation === 'is-new') {
          logger.info(`${logger.underline(destDir)} does not exist yet.`);
          await next('shouldCreateDestDir');
        }
        if (dirValidation === 'is-directory') {
          logger.emptyLine();
          await next('pickConfigs');
        }
      },
    },
    {
      name: 'shouldCreateDestDir',
      message: logger.prompt('Would you like to create it?'),
      type: Toggle,
      after: async ({ shouldCreateDestDir, destDir }, next) => {
        if (!shouldCreateDestDir) {
          logger.emptyLine();
          await next('destDir');
        }
        await Deno.mkdir(destDir as string, { recursive: true });
        hasCreatedDestDir = true;
        await next();
      },
    },
    {
      name: 'pickConfigs',
      type: Checkbox,
      search: true,
      searchLabel: logger.italic('(type to filter ðŸ”Ž)').toString(),
      message: logger.prompt('Select config(s)'),
      options: configFiles,
      maxRows: 25,
      confirmSubmit: false,
      after: async ({ pickConfigs, destDir }, next) => {
        if (!pickConfigs?.length) {
          logger.info("You didn't pick any configs");
          if (hasCreatedDestDir) {
            Deno.remove(normalize(destDir as string), { recursive: true });
          }
        }
        await next();
      },
    },
  ])) satisfies PromptResult;
}
