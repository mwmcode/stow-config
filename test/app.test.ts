import { snapshotTest } from '@cliffy/testing';
import { ansi } from '@cliffy/ansi';
import { runPrompt } from '@/src/runPrompt.ts';
import { DEST_DIR, SRC_DIR, TEST_DENO_ARGS } from './consts.ts';
import { prepareTestData } from './utils.ts';
import { stowConfig } from '@/src/mod.ts';
import { assert, assertExists } from 'jsr:@std/assert';
import { join } from '@std/path';
import { CONFIGS } from '@/test/consts.ts';

const stdInputs = ansi
  .text(`${SRC_DIR}\n`)
  .text(`${DEST_DIR}\n`)
  .text(' ')
  .cursorDown(1)
  .text(' ')
  .cursorDown(1)
  .text(' ')
  .cursorDown(1)
  .text(' \n')
  .toArray();

await snapshotTest({
  name: 'Prompts user to enter paths for source and destination dirs',
  meta: import.meta,
  stdin: stdInputs,
  denoArgs: TEST_DENO_ARGS,
  async fn() {
    await prepareTestData();
    const result = await runPrompt();
    console.log(result);
  },
});

await snapshotTest({
  name: 'Moves config files from src to dest, and creates symlinks',
  meta: import.meta,
  stdin: stdInputs,
  denoArgs: TEST_DENO_ARGS,
  async fn() {
    await prepareTestData();
    const result = await stowConfig();
    console.log(result);
    for (const config of CONFIGS) {
      if (config.includes('/')) {
        const dir = config.split('/')[0];
        assert((await Deno.lstat(join(DEST_DIR, dir))).isDirectory);
        assertExists(await Deno.readLink(join(SRC_DIR, dir)));
        continue;
      }
      assertExists(await Deno.readLink(join(SRC_DIR, config)));
    }
  },
});
