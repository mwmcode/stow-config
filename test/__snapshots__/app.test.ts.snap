export const snapshot = {};

snapshot[`Prompts user to enter paths for source and destination dirs 1`] = `
stdout:
'? Enter path to \`configs\` directory › \\x1b[39G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › test/data\\x1b[47G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › test/data/config-src\\x1b[55G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › /Users/mch/Developer/deno/stow-config/test/data/config-src
\\x1b[?25h\\x1b[?25h\\x1b[?25h

? Enter path to \`destination\` directory › \\x1b[43G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › test/data\\x1b[51G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › test/data/config-src\\x1b[59G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › /Users/mch/Developer/deno/stow-config/test/data/config-dest
\\x1b[?25h\\x1b[?25h\\x1b[?25h

? Select config(s) (type to filter 🔎) 
❯ ✘ .zshrc
  ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
❯ ✔ .zshrc
  ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
❯ ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
❯ ✔ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
❯ ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
❯ ✔ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
  ✔ .prettierrc.js
❯ ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) › .zshrc, starship.toml, .prettierrc.js, git
\\x1b[?25h\\x1b[?25h\\x1b[?25h{
  configDir: "/Users/mch/Developer/deno/stow-config/test/data/config-src",
  destDir: "/Users/mch/Developer/deno/stow-config/test/data/config-dest",
  pickConfigs: [ ".zshrc", "starship.toml", ".prettierrc.js", "git" ]
}
'
stderr:
""
`;

snapshot[`Moves config files from src to dest, and creates symlinks 1`] = `
stdout:
"? Enter path to \`configs\` directory › \\x1b[39G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › test/data\\x1b[47G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › test/data/config-src\\x1b[55G\\x1b[G\\x1b[0J
? Enter path to \`configs\` directory › /Users/mch/Developer/deno/stow-config/test/data/config-src
\\x1b[?25h\\x1b[?25h\\x1b[?25h

? Enter path to \`destination\` directory › \\x1b[43G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › test/data\\x1b[51G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › test/data/config-src\\x1b[59G\\x1b[G\\x1b[0J
? Enter path to \`destination\` directory › /Users/mch/Developer/deno/stow-config/test/data/config-dest
\\x1b[?25h\\x1b[?25h\\x1b[?25h

? Select config(s) (type to filter 🔎) 
❯ ✘ .zshrc
  ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
❯ ✔ .zshrc
  ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
❯ ✘ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
❯ ✔ starship.toml
  ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
❯ ✘ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
❯ ✔ .prettierrc.js
  ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) (type to filter 🔎) 
  ✔ .zshrc
  ✔ starship.toml
  ✔ .prettierrc.js
❯ ✘ git\\x1b[4A\\x1b[40G\\x1b[G\\x1b[0J
? Select config(s) › .zshrc, starship.toml, .prettierrc.js, git
\\x1b[?25h\\x1b[?25h\\x1b[?25h
✅ done!

🔗 /Users/mch/Developer/deno/stow-config/test/data/config-src/.zshrc         @-> /Users/mch/Developer/deno/stow-config/test/data/config-dest/.zshrc        
🔗 /Users/mch/Developer/deno/stow-config/test/data/config-src/starship.toml  @-> /Users/mch/Developer/deno/stow-config/test/data/config-dest/starship.toml 
🔗 /Users/mch/Developer/deno/stow-config/test/data/config-src/.prettierrc.js @-> /Users/mch/Developer/deno/stow-config/test/data/config-dest/.prettierrc.js
🔗 /Users/mch/Developer/deno/stow-config/test/data/config-src/git            @-> /Users/mch/Developer/deno/stow-config/test/data/config-dest/git           
undefined
"
stderr:
""
`;
